### PR TITLE
add comparators guide

### PR DESCRIPTION
**Pull request by josedonizetti**
Hey @puredanger, finally able to get to it.
https://github.com/clojure/clojure-site/issues/15

### CODE REVIEW

**Comment by josedonizetti** [comment]
Let me know what you think. Whenever he used simple code blocks on phrases like `x` you will se a italic  mark on the adoc like _x_. Not sure if the proper way to do it. But I can change it if there's a better form.

**Comment by josedonizetti** [comment]
If this one is good to be merged, is there anything else I can help you with on the clojure-site?

**Comment by puredanger** [comment]
Thanks! 

Do you have a signed CA? I'm not seeing one and can't include the contribution without it. http://clojure.org/contributing

File: content/guides/comparators.adoc [hunk 1]
```diff
@@ -0,0 +1,635 @@
+= Comparators in Clojure
```
Comment (puredanger) [inline content/guides/comparators.adoc]: Let's make this "Comparators Guide"

File: content/guides/comparators.adoc [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+Andy Fingerhut
+2013-04-01
+:jbake-type: page
+:toc: macro
+
+ifdef::env-github,env-browser[:outfilesuffix: .adoc]
+
+== Versions
+
+This document was written while checking the details against Clojure 1.5.1  and Java 6,
```
Comment (puredanger) [inline content/guides/comparators.adoc]: We should write this from the perspective of Clojure 1.7 (the current stable version).

File: content/guides/comparators.adoc [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+A comparator is a function that takes two arguments _x_ and _y_ and returns a value indicating
+the relative order in which _x_ and _y_ should be sorted. It can be a 3-way comparator returning
+an integer, or a 2-way comparator returning a boolean.
+See the DOs below for what the return values should be, depending upon the order of _x_ and _y_.
+
+In Clojure you need comparators for sorting a collection of values, or for maintaining a collection
+of values in a desired sorted order, e.g a
+http://clojure.github.io/clojure/clojure.core-api.html#clojure.core/sorted-map[sorted-map],
+http://clojure.github.io/clojure/clojure.core-api.html#clojure.core/sorted-set[sorted-set], or
+http://clojure.github.io/clojure/clojure.core-api.html#clojure.core/priority-map[priority-map]
```
Comment (puredanger) [inline content/guides/comparators.adoc]: There is no clojure.core/priority-map - this should refer to clojure.data.priority-map (in the contrib library data.priority-map).

File: content/guides/comparators.adoc [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+
+user> (pprint (sort by-salary-name-co people))
+({:name "John", :salary 40000.0, :company "Venus Co"}
+ {:name "John", :salary 35000.0, :company "Acme"}
+ {:name "Mary", :salary 35000.0, :company "Mars Inc"}
+ {:name "John", :salary 30000.0, :company "Asteroids-R-Us"})
+----
+
+Below is the shorter way, by comparing Clojure vectors. It behaves exactly the same as above. Note that
+as above, the field :salary is sorted in descending order becase _x_ and _y_ are swapped.
```
Comment (puredanger) [inline content/guides/comparators.adoc]: "becase" typo

File: content/guides/comparators.adoc [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+----
+
+While this works in many cases, think twice (or three times) before using this technique.
+It is less error-prone to use explicit conditional checks and return -1, 0, or 1, or to use boolean comparators.
+
+Why? Java comparators must return a 32-bit _int_ type, so when a Clojure function is used as a comparator
+and it returns any type of number, that number is converted to an _int_ behind the scenes using the Java method
+http://docs.oracle.com/javase/6/docs/api/java/lang/Number.html#intValue%28%29[intValue].
+See Clojure source file
+https://github.com/clojure/clojure/blob/clojure-1.5.1/src/jvm/clojure/lang/AFunction.java#L46[src/jvm/clojure/lang/AFunction.java]
```
Comment (puredanger) [inline content/guides/comparators.adoc]: use 1.7-equivalent link

File: content/guides/comparators.adoc [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+;; Calling a Clojure function in the normal way uses its invoke
+;; method, not compare.
+user> (. my-< (invoke 2 1))
+(my-< 2 1 ) returns  false
+false
+----
+
+
+See Clojure source file
+https://github.com/clojure/clojure/blob/clojure-1.5.1/src/jvm/clojure/lang/AFunction.java#L46[src/jvm/clojure/lang/AFunction.java]
```
Comment (puredanger) [inline content/guides/comparators.adoc]: use 1.7 link

File: content/guides/comparators.adoc [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+----
+
+== TBD
+
+
+TBD: How does Double/NaN compare to other things?
+The answer is included in the Java docs for sorting an array of doubles:
+http://docs.oracle.com/javase/6/docs/api/java/util/Arrays.html#sort%28double[]
+
+TBD: Maybe mention these other Clojure functions that make use of comparators.
```
Comment (puredanger) [inline content/guides/comparators.adoc]: No, do not mention these - they are deprecated.

File: content/guides/comparators.adoc [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+ 1.7976931348623157E308
+ "b"
+ [1, 2, 3, 4])
+nil
+----
+
+== TBD
+
+
+TBD: How does Double/NaN compare to other things?
```
Comment (puredanger) [inline content/guides/comparators.adoc]: This is not "TBD", it should sort per http://docs.oracle.com/javase/6/docs/api/java/lang/Double.html#compareTo(java.lang.Double), that is it should sort higher than all other doubles, including +Infinity.

File: content/guides/comparators.adoc [hunk 9]
```diff
@@ -0,0 +1,635 @@
...
+The answer is included in the Java docs for sorting an array of doubles:
+http://docs.oracle.com/javase/6/docs/api/java/util/Arrays.html#sort%28double[]
+
+TBD: Maybe mention these other Clojure functions that make use of comparators.
+If mentioned nowhere else, at least at the end of the see also list:
+_clojure.parallel/pmax_, _clojure.parallel/pmin_, _clojure.parallel/psummary_, _clojure.parallel/psort_
+
+TBD: Is function _clojure.core/comparator_ useful for anything?
```
Comment (puredanger) [inline content/guides/comparators.adoc]: Yes, there are uses for it - remove this.

**Comment by puredanger** [comment]
I'm done reviewing - if you sign a CA and make the changes above, I will merge. Thanks!

**Comment by josedonizetti** [comment]
Changes done. and CA signed. Thank you.

**Comment by puredanger** [comment]
New guide is merged - I undid the index changes. The new guide will be integrated in a different way when the production style and templates are merged.

**Comment by josedonizetti** [comment]
Sure! Anything else I can help with on the clojure-site?

**Pull request merged**